warp10/src/main/java/io/warp10/script/functions/FETCH.java

```diff
@@ -1099,4 +984,197 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
+            }
+          } catch (WarpScriptException | IllegalArgumentException e) {
+            throw new WarpScriptException("Invalid format for parameter '" + startParamName + "'.", e);
+          }
+        }
+      }
+    } else if (null != start) {
+      // If start is not null and we cannot retrieve the timestamp, throw an error.
+      throw new WarpScriptException("Invalid format for parameter '" + startParamName + "'.");
+    }
+
+    //
+    // Try to convert end to a valid timestamp if possible.
+    // Same exact logic as for start.
+    //
+
+    Long endTs = null;
+
+    if (end instanceof Long) {
+      // Simple case: start is a long
+      endTs = (long) end;
+    } else if (end instanceof String) {
+      // If it's a string, it may be the string representation of a Long, a ISO8601 date or 'now'.
+      if ("now".equals(start)) {
+        endTs = now;
+      } else {
+        try {
+          endTs = Long.valueOf((String) end);
+        } catch (NumberFormatException nfe) {
+          // Not string representation of a Long, try ISO8601
+          try {
+            if (SystemUtils.isJavaVersionAtLeast(JavaVersion.JAVA_1_8)) {
+              endTs = io.warp10.script.unary.TOTIMESTAMP.parseTimestamp((String) end);
+            } else {
+              endTs = fmt.parseDateTime((String) end).getMillis() * Constants.TIME_UNITS_PER_MS;
+            }
+          } catch (WarpScriptException | IllegalArgumentException e) {
+            throw new WarpScriptException("Invalid format for parameter '" + endParamName + "'.", e);
+          }
+        }
+      }
+    } else if (null != end) {
+      // If end is not null and we cannot retrieve the timestamp, throw an error.
+      throw new WarpScriptException("Invalid format for parameter '" + endParamName + "'.");
+    }
+
+    // Check that either startTs or endTs is defined.
+    if (null == startTs && null == endTs) {
+      throw new WarpScriptException("Missing either '" + startParamName + "' or '" + endParamName + "' parameter.");
+    }
+
+    // If both are defined but swapped, swap them.
+    if (null != startTs && null != endTs && startTs > endTs) {
+      long tmp = startTs;
+      startTs = endTs;
+      endTs = tmp;
+    }
+
+    //
+    // Try to use timestamp to either determine endTs or startTs if not already defined.
+    //
+
+    Long numericTimespan = null;
+
+    if (null != timespan) {
+      // Check that endTs and startTs are not both already defined.
+      if (startTs != null && endTs != null) {
+        throw new WarpScriptException("Invalid time range specification: '" + startParamName + "', '" + endParamName + "' and '" + timespanParamName + "' cannot all be defined. Only 2 out of those 3 parameters should be defined.");
+      }
+
+      //
+      // Cast or convert timespan to a numeric one.
+      //
+      if (timespan instanceof Long) {
+        numericTimespan = (Long) timespan;
```

Comment:
Timespan is assumed to be >= 0 in this method, that is because the calling path you set up ensures so, but maybe this should be checked here and after line 1106.